`@mroysterhead ➜ /workspaces/learn-to-code-go-version-03/dice_roller (dice_roller) $ go test
main_test.go:1:1: expected 'package', found 'func'`

I think I am going to dial back the dis adv feature and get a working script with a working test file before building up more stuff

--- a/dice_roller/main.go
+++ b/dice_roller/main.go
@@ -1,7 +1,6 @@
 package main
 
 import (
-	"flag"
 	"fmt"
 	"math/rand"
 	"regexp"
@@ -9,81 +8,41 @@ import (
 	"time"
 )
 
-func parseInput(input string) (int, int, int, error) {
-	diceExpr := regexp.MustCompile(`^(\d+)d(\d+)([+-]\d+)?$`)
-	matches := diceExpr.FindStringSubmatch(input)
-
-	if matches == nil {
-		return 0, 0, 0, fmt.Errorf("invalid input format")
-	}
-
-	numDice, err := strconv.Atoi(matches[1])
-	if err != nil {
-		return 0, 0, 0, err
-	}
-
-	numSides, err := strconv.Atoi(matches[2])
-	if err != nil {
-		return 0, 0, 0, err
-	}
-
-	modifier := 0
-	if len(matches) > 3 {
-		modifier, err = strconv.Atoi(matches[3])
-		if err != nil {
-			return 0, 0, 0, err
-		}
-	}
-
-	return numDice, numSides, modifier, nil
-}
-
-func rollDice(numDice, numSides int, adv, dis *bool) int {
-	total := 0
-	for i := 0; i < numDice; i++ {
-		roll := rand.Intn(numSides) + 1
-		if *adv {
-			roll = max(roll, rand.Intn(numSides)+1)
-		} else if *dis {
-			roll = min(roll, rand.Intn(numSides)+1)
-		}
-		fmt.Printf("Die %d: %d\n", i+1, roll)
-		total += roll
-	}
-	return total
-}
-
 func main() {
-	adv := flag.Bool("adv", false, "Roll twice and take the higher")
-	dis := flag.Bool("dis", false, "Roll twice and take the lower")
-	flag.Parse()
+	// Regular expression to validate input format (XdY[+-]N)
+	diceExpr := regexp.MustCompile(`^(\d+)d(\d+)([+-]\d+)?$`)
 
-	fmt.Println("Enter dice expression ([number of dice]d[size of dice]+/-[modifier] ie:1d20+4): ")
+	fmt.Println("Enter dice expression (XdY[+-]N): ")
 	var input string
 	fmt.Scanln(&input)
 
-	numDice, numSides, modifier, err := parseInput(input)
-	if err != nil {
-		fmt.Println(err)
+	// Validate input format
+	if !diceExpr.MatchString(input) {
+		fmt.Println("Invalid input format. Please enter XdY[+-]N where X is the number of dice, Y is the number of sides, and N is the optional modifier.")
 		return
 	}
 
-	rand.Seed(time.Now().UnixNano())
-	total := rollDice(numDice, numSides, adv, dis)
+	// Extract number of dice, sides, and modifier (if present) from input
+	matches := diceExpr.FindStringSubmatch(input)
+	numDice, _ := strconv.Atoi(matches[1])
+	numSides, _ := strconv.Atoi(matches[2])
+	modifier := 0
+	if len(matches) > 3 {
+		modifier, _ = strconv.Atoi(matches[3])
+	}
+
+	// Seed random number generator
+	rand.Seed(int64(time.Now().UnixNano()))
+
+	// Roll dice and sum results
+	var total int
+	for i := 0; i < numDice; i++ {
+		roll := rand.Intn(numSides) + 1
+		fmt.Printf("Die %d: %d\n", i+1, roll)
+		total += roll
+	}
+
+	// Apply modifier and print total
 	total += modifier
 	fmt.Printf("Total: %d (including modifier of %d)\n", total, modifier)
-}
-
-func max(a, b int) int {
-	if a > b {
-		return a
-	}
-	return b
-}
-
-func min(a, b int) int {
-	if a < b {
-		return a
-	}
-	return b
 }

--- a/dice_roller/main_test.go
+++ b/dice_roller/main_test.go
@@ -1,32 +1,46 @@
 package main
 
-import "testing"
+import (
+	"io/ioutil"
+	"os"
+	"strconv"
+	"strings"
+	"testing"
+)
 
-func TestParseInput(t *testing.T) {
+func TestRollDice(t *testing.T) {
 	tests := []struct {
-		input                                               string
-		expectedNumDice, expectedNumSides, expectedModifier int
-		expectErr                                           bool
+		input    string
+		expected int
 	}{
-		{"3d6+1", 3, 6, 1, false},
-		{"2d20-5", 2, 20, -5, false},
-		{"5d10", 5, 10, 0, false},
-		{"invalid", 0, 0, 0, true},
+		{"2d6", 7},     // Simple roll
+		{"1d20+5", 18}, // Roll with modifier
+		{"4d4-2", 9},   // Multiple dice with modifier
 	}
 
 	for _, test := range tests {
-		numDice, numSides, modifier, err := parseInput(test.input)
-		if (err != nil) != test.expectErr {
-			t.Errorf("parseInput(%q) error = %v, expectErr %v", test.input, err, test.expectErr)
-			continue
-		}
-		if numDice != test.expectedNumDice || numSides != test.expectedNumSides || modifier != test.expectedModifier {
-			t.Errorf("parseInput(%q) = %v, %v, %v, want %v, %v, %v", test.input, numDice, numSides, modifier, test.expectedNumDice, test.expectedNumSides, test.expectedModifier)
+		// Capture standard output to compare results
+		old := os.Stdout
+		r, w, _ := os.Pipe()
+		os.Stdout = w
+
+		// Call the main function with the test input
+		main()
+
+		// Restore standard output
+		w.Close()
+		os.Stdout = old
+
+		// Read the captured output
+		output, _ := ioutil.ReadAll(r)
+		resultLine := strings.Split(string(output), "\n")[len(strings.Split(string(output), "\n"))-1]
+		splitResult := strings.Split(resultLine, ": ")
+		if len(splitResult) >= 2 {
+			actual, _ := strconv.Atoi(splitResult[1])
+			// Use actual here
+			if actual != test.expected {
+				t.Errorf("Roll failed for input %s: expected %d, got %d", test.input, test.expected, actual)
+			}
 		}
 	}
-}
-
-func parseInput(s string) (numDice, numSides, modifier int, err error) {
-	// Your code here. Don't forget to return the appropriate values at the end.
-	return 0, 0, 0, nil // Placeholder return statement. Replace with your actual implementation.
 }

--- a/dice_roller/main_test.go
+++ b/dice_roller/main_test.go
@@ -1,3 +1,7 @@
+package main
+
+import "testing"
+
 func TestParseInput(t *testing.T) {
 	tests := []struct {
 		input                                               string
@@ -20,4 +24,9 @@ func TestParseInput(t *testing.T) {
 			t.Errorf("parseInput(%q) = %v, %v, %v, want %v, %v, %v", test.input, numDice, numSides, modifier, test.expectedNumDice, test.expectedNumSides, test.expectedModifier)
 		}
 	}
+}
+
+func parseInput(s string) (numDice, numSides, modifier int, err error) {
+	// Your code here. Don't forget to return the appropriate values at the end.
+	return 0, 0, 0, nil // Placeholder return statement. Replace with your actual implementation.
 }

